File: core/concurrent/java/fan/concurrent/AtomicBoolPeer.java
//
// Copyright (c) 2010, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Feb 10 Brian Frank  Creation
//
package fan.concurrent;

import java.util.concurrent.atomic.AtomicBoolean;

public final class AtomicBoolPeer
{

  public static AtomicBoolPeer make(AtomicBool self)
  {
    return new AtomicBoolPeer();
  }

  public final boolean val(AtomicBool self)
  {
    return val.get();
  }

  public final void val(AtomicBool self, boolean v)
  {
    val.set(v);
  }

  public final boolean getAndSet(AtomicBool self, boolean v)
  {
    return val.getAndSet(v);
  }

  public final boolean compareAndSet(AtomicBool self, boolean expect, boolean update)
  {
    return val.compareAndSet(expect, update);
  }

  private final AtomicBoolean val = new AtomicBoolean();
}

File: src/concurrent/js/AtomicBoolPeer.js
//
// Copyright (c) 2015, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   17 Aug 2015  Matthew Giannini  Creation
//

fan.concurrent.AtomicBoolPeer = fan.sys.Obj.$extend(fan.sys.Obj);
fan.concurrent.AtomicBoolPeer.prototype.$ctor = function() {}

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

fan.concurrent.AtomicBoolPeer.prototype.m_val = false;
fan.concurrent.AtomicBoolPeer.prototype.val = function(self) { return this.m_val; }
fan.concurrent.AtomicBoolPeer.prototype.val$ = function(self, val) { this.m_val = val; }

fan.concurrent.AtomicBoolPeer.prototype.getAndSet = function(self, val) {
  var old = this.m_val;
  this.m_val = val;
  return old;
}

fan.concurrent.AtomicBoolPeer.prototype.compareAndSet = function(self, expect, update) {
  if (this.m_val == expect) {
    this.m_val = update;
    return true;
  }
  return false;
}
